### PR TITLE
[Merged by Bors] - feat : added `eigenvalues_mem_spectrum_real` and supporting RCLike coercion results

### PR DESCRIPTION
**Pull request by JonBannon**
Added several RCLike to real coercion results to support the result that eigenvalues of a hermitian matrix belong to the real spectrum of that matrix, and included this result.

Co-Authored by @j-loreaux

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 1]
```diff
@@ -55,6 +56,33 @@ lemma mulVec_eigenvectorBasis (j : n) :
       congr(⇑$((isHermitian_iff_isSymmetric.1 hA).apply_eigenvectorBasis
         finrank_euclideanSpace ((Fintype.equivOfCardEq (Fintype.card_fin _)).symm j)))
 
+/--Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
+theorem eigenvalue_mem_toEuclideanLin_spectrum_RCLike (i : n) :
+    (RCLike.ofReal ∘ hA.eigenvalues) i ∈ spectrum 𝕜 (toEuclideanLin A) :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:61]: I think this should be written as
```suggestion
theorem ofReal_eigenvalue_mem_spectrum_toEuclideanLin (i : n) :
    hA.eigenvalues i ∈ spectrum 𝕜 (toEuclideanLin A) :=
```
where hopefully the coercion is inserted automatically

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 2]
```diff
@@ -55,6 +56,33 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+  LinearMap.IsSymmetric.hasEigenvalue_eigenvalues _ _ _ |>.mem_spectrum
+
+/-- Algebra equivalence between the linear maps and continuous linear maps on a finite-dim module.-/
+def AlgEquivFiniteDimNormedLinearCLM.{v} (E : Type v) [NormedAddCommGroup E]
+    [NormedSpace 𝕜 E][FiniteDimensional 𝕜 E] :
+    AlgEquiv (R := 𝕜) (A := E →ₗ[𝕜] E) (B := E →L[𝕜] E) :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:67]: Please use the notation for `AlgEquiv` here.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 3]
```diff
@@ -55,6 +56,33 @@ lemma mulVec_eigenvectorBasis (j : n) :
       congr(⇑$((isHermitian_iff_isSymmetric.1 hA).apply_eigenvectorBasis
         finrank_euclideanSpace ((Fintype.equivOfCardEq (Fintype.card_fin _)).symm j)))
 
+/--Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
+theorem eigenvalue_mem_toEuclideanLin_spectrum_RCLike (i : n) :
+    (RCLike.ofReal ∘ hA.eigenvalues) i ∈ spectrum 𝕜 (toEuclideanLin A) :=
+  LinearMap.IsSymmetric.hasEigenvalue_eigenvalues _ _ _ |>.mem_spectrum
+
+/-- Algebra equivalence between the linear maps and continuous linear maps on a finite-dim module.-/
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:64]: Please mention `LinearMap.toContinuousLinearMap` in the docstring as a cross-reference.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 4]
```diff
@@ -55,6 +56,33 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+def AlgEquivFiniteDimNormedLinearCLM.{v} (E : Type v) [NormedAddCommGroup E]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:65]: Perhaps this should be called `Module.End.toContinuousLinearMap` to match?

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 5]
```diff
@@ -55,6 +56,33 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+/--Spectrum of a Hermitian matrix equals the spectrum as a EuclideanLin. -/
+theorem spec_toEuclideanLin_eq_spec : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
+  AlgEquiv.spectrum_eq ((AlgEquiv.trans ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) : Matrix n n 𝕜 ≃ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n))
+  (AlgEquivFiniteDimNormedLinearCLM (EuclideanSpace 𝕜 n)).symm) _
+
+/--Eigenvalues of a hermitian matrix A are in the ℝ spectrum of A. -/
+theorem eigenvalue_mem_real : ∀ (i : n), (hA.eigenvalues) i ∈ spectrum ℝ A := by
+  intro i
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:81]:
```suggestion
theorem eigenvalues_mem_spectrum_real (i : n) : hA.eigenvalues i ∈ spectrum ℝ A := by
```

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 1 above)
Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:61]: Implemented name change, but unfortunately if I remove the explicit coercion, the result won't compile.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 4 above)
Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:65]: Thanks for all the helpful suggestions! Should this result be moved into Algebra.Module._ , or do you think it still fits here?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean]: This result should probably move next to `LinearMap.toContinuousLinearMap` if possible. If not, `#find_home` might have some useful suggestions as to where to put it. Certainly it should not need `RCLike`, as `LinearMap.toContinuousLinearMap` doesn't.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean]: I pushed a fix

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 6]
```diff
@@ -55,6 +56,32 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+  __ := LinearMap.toContinuousLinearMap
+  map_mul' _ _ := rfl
+  commutes' _ := rfl
+
+/--Spectrum of a Hermitian matrix equals the spectrum as a EuclideanLin. -/
+theorem spec_toEuclideanLin_eq_spec : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
+  AlgEquiv.spectrum_eq ((AlgEquiv.trans ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) : Matrix n n 𝕜 ≃ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n))
+  (Module.End.toContinuousLinearMap (EuclideanSpace 𝕜 n)).symm) _
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:77]: I find this very hard to parse in my head. Does this work?
```suggestion
  AlgEquiv.spectrum_eq
    (AlgEquiv.trans
      ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜] EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) :
          Matrix n n 𝕜 ≃ₐ[𝕜] EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n)
      (Module.End.toContinuousLinearMap (EuclideanSpace 𝕜 n)).symm)
    _
```

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 7]
```diff
@@ -55,6 +56,32 @@ lemma mulVec_eigenvectorBasis (j : n) :
       congr(⇑$((isHermitian_iff_isSymmetric.1 hA).apply_eigenvectorBasis
         finrank_euclideanSpace ((Fintype.equivOfCardEq (Fintype.card_fin _)).symm j)))
 
+/--Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:59]: Is this what you meant?
```suggestion
/-- Eigenvalues of a Hermitian matrix, coerced, belong to the spectrum of the associated
`toEuclideanLin`. -/
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
Thanks!

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 6 above)
Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:77]: Let me try it...

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 7 above)
Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean]: Yes, thanks!

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 6 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:78]: Yes, I think this is a style change: the idea is to reflect parenthesis nesting with indentation.

**Comment by JonBannon** [comment]
Ah, ok. I'm happy to commit the suggestion if it confirms to style
guidelines!

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 8]
```diff
@@ -55,6 +57,24 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:60]:
```suggestion
/-- Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
```

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 9]
```diff
@@ -55,6 +57,24 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+theorem ofReal_eigenvalue_mem_spectrum_toEuclideanLin (i : n) :
+    (RCLike.ofReal ∘ hA.eigenvalues) i ∈ spectrum 𝕜 (toEuclideanLin A) :=
+  LinearMap.IsSymmetric.hasEigenvalue_eigenvalues _ _ _ |>.mem_spectrum
+
+/--Spectrum of a Hermitian matrix equals the spectrum as a EuclideanLin. -/
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:65]:
```suggestion
/-- The spectrum of a Hermitian matrix `A` coincides with the spectrum of `toEuclideanLin A`. -/
```

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 6 above)
Comment (j-loreaux) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:70]: It doesn't look like you ever made this change

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 10]
```diff
@@ -55,6 +57,24 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+theorem spec_toEuclideanLin_eq_spec : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:66]: we don't abbreviate `spectrum` to `spec` anywhere else in the library.
```suggestion
theorem spectrum_toEuclideanLin : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
```

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 11]
```diff
@@ -55,6 +57,24 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+  AlgEquiv.spectrum_eq ((AlgEquiv.trans ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) : Matrix n n 𝕜 ≃ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n))
+  (Module.End.toContinuousLinearMap (EuclideanSpace 𝕜 n)).symm) _
+
+/--Eigenvalues of a hermitian matrix A are in the ℝ spectrum of A. -/
+theorem eigenvalues_mem_spectrum_real (i : n) : hA.eigenvalues i ∈ spectrum ℝ A := by
+  apply spectrum.of_algebraMap_mem (S := 𝕜) (R := ℝ) (A := Matrix n n 𝕜)
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:74]: Does this not work?
```suggestion
  apply spectrum.of_algebraMap_mem 𝕜
```

File: Mathlib/Topology/Algebra/Module/FiniteDimension.lean [hunk 12]
```diff
@@ -290,6 +290,14 @@ def toContinuousLinearMap : (E →ₗ[𝕜] F') ≃ₗ[𝕜] E →L[𝕜] F' whe
   right_inv _ := ContinuousLinearMap.coe_injective rfl
 #align linear_map.to_continuous_linear_map LinearMap.toContinuousLinearMap
 
+/-- Algebra equivalence between the linear maps and continuous linear maps on a finite dimensional
+    space. -/
+def _root_.Module.End.toContinuousLinearMap (E : Type v) [NormedAddCommGroup E]
+    [NormedSpace 𝕜 E][FiniteDimensional 𝕜 E] : (E →ₗ[𝕜] E) ≃ₐ[𝕜] (E →L[𝕜] E) :=
+    {LinearMap.toContinuousLinearMap with
+    map_mul' := fun _ _ ↦ rfl
+    commutes' := fun _ ↦ rfl}
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/Module/FiniteDimension.lean:299]:
```suggestion
    [NormedSpace 𝕜 E] [FiniteDimensional 𝕜 E] : (E →ₗ[𝕜] E) ≃ₐ[𝕜] (E →L[𝕜] E) :=
  { LinearMap.toContinuousLinearMap with
    map_mul' := fun _ _ ↦ rfl
    commutes' := fun _ ↦ rfl }
```

**Review by j-loreaux — COMMENTED** [review]
style, naming, and some simplification

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean (hunk 11 above)
Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:76]: Will check. (I hope that it will!)

Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:76]: Yep.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 13]
```diff
@@ -55,6 +57,26 @@ lemma mulVec_eigenvectorBasis (j : n) :
       congr(⇑$((isHermitian_iff_isSymmetric.1 hA).apply_eigenvectorBasis
         finrank_euclideanSpace ((Fintype.equivOfCardEq (Fintype.card_fin _)).symm j)))
 
+/-- Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
+theorem ofReal_eigenvalue_mem_spectrum_toEuclideanLin (i : n) :
+    (RCLike.ofReal ∘ hA.eigenvalues) i ∈ spectrum 𝕜 (toEuclideanLin A) :=
+  LinearMap.IsSymmetric.hasEigenvalue_eigenvalues _ _ _ |>.mem_spectrum
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:63]: How do you feel about just inlining this in the proof of `eigenvalues_mem_spectrum_real` below? We can always get back easily with `spectrum.algebraMap_mem` and `spectrum_toEuclideanLin`. I just feel like we don't really need many results about `spectrum 𝕜 (toEuclideanLin A)`.

Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:63]: This is totally fine...I feel a bit weird about *having* these coercion-involved results, so inlining them seems better. Not sure this is the correct intuition, but...trying to learn the right design mojo...

Comment (JonBannon) [reply Mathlib/LinearAlgebra/Matrix/Spectrum.lean:63]: Used apply, but I think it's probably ok.

File: Mathlib/LinearAlgebra/Matrix/Spectrum.lean [hunk 14]
```diff
@@ -55,6 +57,21 @@ lemma mulVec_eigenvectorBasis (j : n) :
...
+/-- The spectrum of a Hermitian matrix `A` coincides with the spectrum of `toEuclideanLin A`. -/
+theorem spectrum_toEuclideanLin : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
+  AlgEquiv.spectrum_eq
+    (AlgEquiv.trans
+      ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜] EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) :
+          Matrix n n 𝕜 ≃ₐ[𝕜] EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n)
+      (Module.End.toContinuousLinearMap (EuclideanSpace 𝕜 n)).symm)
+    _
+
+/--Eigenvalues of a hermitian matrix A are in the ℝ spectrum of A. -/
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/Spectrum.lean:69]:
```suggestion
/-- Eigenvalues of a hermitian matrix A are in the ℝ spectrum of A. -/
```

**Comment by j-loreaux** [comment]
I've had a lot of input here, so

maintainer merge

**Comment by dupuisf** [comment]
Thanks!

bors r+
